File: translate-web-app/backend/worker.js
const Bull = require('bull');
const path = require('path');
const fs = require('fs').promises;
const DocumentTranslator = require('./translator');
const { downloadFromS3, uploadToS3 } = require('./s3Helper');
// Load .env from project root (parent of translate-web-app)
require('dotenv').config({ path: path.join(__dirname, '../../.env') });

// Redis configuration from environment variable
const redisUrl = process.env.REDIS_URL;

if (!redisUrl) {
  console.error('Error: REDIS_URL not found in environment variables');
  process.exit(1);
}

// Parse Redis URL manually for Bull compatibility
const parseRedisUrl = (url) => {
  const urlObj = new URL(url);
  return {
    host: urlObj.hostname,
    port: parseInt(urlObj.port, 10),
    password: urlObj.password || undefined,
    db: urlObj.pathname ? parseInt(urlObj.pathname.slice(1), 10) || 0 : 0,
    tls: urlObj.protocol === 'rediss:' ? { rejectUnauthorized: false } : undefined
  };
};

const redisOptions = parseRedisUrl(redisUrl);
console.log('Worker Redis config:', { host: redisOptions.host, port: redisOptions.port, hasTLS: !!redisOptions.tls });

const redisConfig = {
  redis: redisOptions,
  defaultJobOptions: {
    removeOnComplete: 100,
    removeOnFail: 100,
    attempts: 3,
    backoff: {
      type: 'exponential',
      delay: 2000
    }
  }
};

// Create queue
const translationQueue = new Bull('translation-queue', redisConfig);

// Initialize translator
const translator = new DocumentTranslator();

// Worker configuration
const WORKER_CONCURRENCY = parseInt(process.env.WORKER_CONCURRENCY) || 5;

// Check dependencies on startup
(async () => {
  try {
    console.log('Checking Python dependencies...');
    await translator.checkPythonDependencies();
    console.log('All dependencies ready');
  } catch (error) {
    console.error('Failed to initialize:', error.message);
    process.exit(1);
  }
})();

// Process translation jobs
translationQueue.process('translate-document', WORKER_CONCURRENCY, async (job) => {
  const {
    fileName,
    originalName,
    s3Key,
    targetLanguage,
    sourceLanguage,
    userId
  } = job.data;

  console.log(`Processing job ${job.id} - File: ${originalName}`);
  console.log(`S3 Key: ${s3Key}`);

  // Create temp directories
  const tempDir = path.join(__dirname, 'temp');
  await fs.mkdir(tempDir, { recursive: true });

  const localInputPath = path.join(tempDir, `input_${job.id}_${fileName}`);
  const localOutputPath = path.join(tempDir, `output_${job.id}_${fileName}`);

  try {
    // Report initial progress
    await job.progress(5);

    // Download file from S3 to local temp
    console.log(`Downloading from S3: ${s3Key}`);
    await downloadFromS3(s3Key, localInputPath);
    console.log(`Downloaded to: ${localInputPath}`);

    // Report progress
    await job.progress(10);

    // Perform translation
    console.log(`Translating ${originalName} to ${targetLanguage}...`);

    const translationResult = await translator.translateDocument(
      localInputPath,
      localOutputPath,
      targetLanguage,
      process.env.GEMINI_API_KEY,
      async (progress) => {
        // Map translation progress (10-85%)
        const mappedProgress = Math.min(85, 10 + Math.ceil(progress * 75));
        await job.progress(mappedProgress);
      }
    );

    // Use actual output path from Python (for PDFs) or the expected path
    let actualOutputPath = localOutputPath;
    let actualOutputFileName = path.basename(localOutputPath);

    if (translationResult.paths) {
      if (translationResult.paths.mono) {
        // PDF translation returns mono and dual paths
        actualOutputPath = translationResult.paths.mono;
        actualOutputFileName = path.basename(actualOutputPath);
        console.log(`PDF translation complete. Mono: ${actualOutputPath}`);
        console.log(`Dual version: ${translationResult.paths.dual}`);
      } else if (translationResult.paths.output) {
        actualOutputPath = translationResult.paths.output;
        actualOutputFileName = path.basename(actualOutputPath);
      }
    }

    // Verify output file exists
    await fs.access(actualOutputPath);

    // Upload result to S3
    await job.progress(90);
    const outputS3Key = `outputs/${actualOutputFileName}`;
    console.log(`Uploading result to S3: ${outputS3Key}`);
    await uploadToS3(actualOutputPath, outputS3Key);
    console.log(`Result uploaded to S3: ${outputS3Key}`);

    // Handle dual path for PDFs
    if (translationResult.paths && translationResult.paths.dual) {
      const dualFileName = path.basename(translationResult.paths.dual);
      const dualS3Key = `outputs/${dualFileName}`;
      console.log(`Uploading dual version to S3: ${dualS3Key}`);
      await uploadToS3(translationResult.paths.dual, dualS3Key);
    }

    // Clean up local temp files
    try {
      await fs.unlink(localInputPath);
      await fs.unlink(actualOutputPath);
      if (translationResult.paths && translationResult.paths.dual) {
        await fs.unlink(translationResult.paths.dual);
      }
    } catch (error) {
      console.error('Error cleaning up temp files:', error);
    }

    // Final progress
    await job.progress(100);

    console.log(`Job ${job.id} completed successfully`);

    const result = {
      success: true,
      outputFile: actualOutputFileName,
      outputPath: actualOutputPath,
      originalName: originalName,
      targetLanguage: targetLanguage,
      completedAt: new Date().toISOString()
    };

    // Add dual path for PDFs if available
    if (translationResult.paths && translationResult.paths.dual) {
      result.dualOutputPath = translationResult.paths.dual;
      result.dualOutputFile = path.basename(translationResult.paths.dual);
    }

    return result;

  } catch (error) {
    console.error(`Job ${job.id} failed:`, error);

    // Clean up temp files on error
    try {
      await fs.unlink(localInputPath);
    } catch (cleanupError) {
      console.error('Cleanup error:', cleanupError);
    }
    try {
      await fs.unlink(localOutputPath);
    } catch (cleanupError) {
      // Ignore if output doesn't exist
    }

    throw error;
  }
});

// Queue event handlers
translationQueue.on('error', (error) => {
  console.error('Queue error:', error);
});

translationQueue.on('stalled', (job) => {
  console.warn(`Job ${job.id} stalled and will be retried`);
});

translationQueue.on('completed', (job, result) => {
  console.log(`Job ${job.id} completed:`, result.outputFile);
});

translationQueue.on('failed', (job, error) => {
  console.error(`Job ${job.id} failed:`, error.message);
});

// Graceful shutdown
process.on('SIGTERM', async () => {
  console.log('SIGTERM received, closing queue...');
  await translationQueue.close();
  process.exit(0);
});

process.on('SIGINT', async () => {
  console.log('SIGINT received, closing queue...');
  await translationQueue.close();
  process.exit(0);
});

console.log(`Worker started with concurrency: ${WORKER_CONCURRENCY}`);
console.log('Waiting for jobs...');

File: translate-web-app/backend/worker.py
"""
RQ Worker for processing translation jobs from Redis queue
"""

import os
import sys
from pathlib import Path

# Load environment variables FIRST
from dotenv import load_dotenv
env_path = Path(__file__).parent.parent.parent / '.env'
load_dotenv(env_path)

# Explicitly set GEMINI_API_KEY environment variable
GEMINI_API_KEY = os.environ.get('GEMINI_API_KEY')
if not GEMINI_API_KEY:
    # Try loading it directly from .env file
    if env_path.exists():
        with open(env_path) as f:
            for line in f:
                if line.startswith('GEMINI_API_KEY='):
                    GEMINI_API_KEY = line.split('=', 1)[1].strip()
                    os.environ['GEMINI_API_KEY'] = GEMINI_API_KEY
                    break
    if not GEMINI_API_KEY:
        raise ValueError("GEMINI_API_KEY environment variable is required")

# Add parent directory to path to import translators
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from translators import DocumentTranslator
from s3_helper import get_s3_helper
from rq import get_current_job

# Initialize translator
MAX_WORKERS = int(os.environ.get('MAX_WORKERS', '4'))
translator = DocumentTranslator(
    api_key=GEMINI_API_KEY,
    model="gemini-2.0-flash-lite",
    max_workers=MAX_WORKERS
)

# Initialize S3 helper
s3_helper = get_s3_helper()


def process_translation(s3_key: str, filename: str, target_language: str):
    """
    Process translation job (S3-based for multi-dyno support)

    Args:
        s3_key: S3 key of uploaded file
        filename: Original filename
        target_language: Target language for translation

    Returns:
        str: Output S3 key on success
    """
    job = get_current_job()

    try:
        if not s3_helper:
            raise Exception("S3 not configured - cannot process job")

        # Update job metadata
        job.meta['progress'] = 5
        job.meta['message'] = 'Downloading file from storage...'
        job.save_meta()

        # Download from S3 to temp location
        import tempfile
        temp_dir = tempfile.mkdtemp()
        input_path = Path(temp_dir) / filename

        download_success = s3_helper.download_file(s3_key, str(input_path))
        if not download_success:
            raise Exception(f"Failed to download file from S3: {s3_key}")

        # Determine output path
        output_filename = f"{input_path.stem}_translated_{target_language}{input_path.suffix}"
        output_path = Path(temp_dir) / output_filename

        # Progress callback
        def progress_callback(progress: float):
            """Update job progress"""
            percent = int(progress * 100)
            job.meta['progress'] = percent
            job.meta['message'] = f"Translating... {percent}%"
            job.save_meta()

        # Perform translation - ensure all paths are strings
        result = translator.translate_document(
            input_path=str(input_path.resolve()),  # Absolute path as string
            output_path=str(output_path.resolve()),  # Absolute path as string
            target_language=target_language,
            progress_callback=progress_callback
        )

        if result:
            # Success
            # For PDFs, result is a tuple (mono_path, dual_path), for others it's True/False
            if isinstance(result, tuple) and len(result) == 2:
                # PDF translation - use the mono path
                mono_path, dual_path = result
                actual_output_path = Path(mono_path)
                actual_output_filename = actual_output_path.name

                # Upload to S3
                output_s3_key = f"outputs/{actual_output_filename}"
                upload_success = s3_helper.upload_file(str(actual_output_path), output_s3_key)

                if not upload_success:
                    raise Exception("Failed to upload result to S3")

                job.meta['progress'] = 100
                job.meta['message'] = 'Translation completed successfully'
                job.save_meta()

                return actual_output_filename
            else:
                # Non-PDF translation
                output_s3_key = f"outputs/{output_filename}"
                upload_success = s3_helper.upload_file(str(output_path), output_s3_key)

                if not upload_success:
                    raise Exception("Failed to upload result to S3")

                job.meta['progress'] = 100
                job.meta['message'] = 'Translation completed successfully'
                job.save_meta()

                return output_filename
        else:
            raise Exception("Translation failed")

    except Exception as e:
        # Failed
        job.meta['progress'] = 0
        job.meta['message'] = 'Translation failed'
        job.meta['error'] = str(e)
        job.save_meta()
        raise
    finally:
        # Clean up temp files
        try:
            import shutil
            if 'temp_dir' in locals():
                shutil.rmtree(temp_dir, ignore_errors=True)
        except:
            pass

        # Clean up uploaded file from S3
        try:
            if 's3_key' in locals() and s3_helper:
                s3_helper.delete_file(s3_key)
        except:
            pass


if __name__ == '__main__':
    # Run worker (can be started with: rq worker translation)
    from redis import Redis
    from rq import Worker

    REDIS_URL = os.environ.get('REDIS_URL')
    if not REDIS_URL:
        raise ValueError("REDIS_URL environment variable is required")

    redis_conn = Redis.from_url(
        REDIS_URL,
        decode_responses=False,
        ssl_cert_reqs=None  # Disable SSL certificate verification
    )

    # Create worker
    worker = Worker(['translation'], connection=redis_conn)

    print("Worker started. Listening for translation jobs...")
    worker.work()
